src/compiler/builderState.ts

```diff
@@ -121,26 +132,33 @@ namespace ts {
 
                         set.forEach(v => deleteFromMultimap(reverse, v, k));
                         forward.delete(k);
+                        version++;
                         return true;
                     },
                     set: (k, vSet) => {
-                        deleted?.delete(k);
+                        let changed = !!deleted?.delete(k);
 
                         const existingVSet = forward.get(k);
                         forward.set(k, vSet);

```

Comment:
So if this actually results in a change, it will necessarily get hit in one of the following two loops, right? (And if `existingVSet` is `undefined` it will be set to `true` in the second loop?) Is it possible to ever see an empty set? It looks to me like the only way we could make a mutation without setting `changed` is by setting an empty `vSet` for a `k` that did not exist before. Assuming that can’t happen and wouldn’t be semantically meaningful if it did, this looks right to me.

Reply:
Yes, AFAIK, consuming the return value of `delete` is [redundant](https://github.com/microsoft/TypeScript/pull/45238#discussion_r679388151).

`deleteFromMultimap` clears empty sets by default, so I believe an empty set is impossible.  Also, if it did happen, I don't think I would count that as a change, since it wouldn't actually add a new mapping (which I think is what you meant by "wouldn't be semantically meaningful").